### PR TITLE
[MODULAR] Nanite Slurry Enhancements

### PR DESCRIPTION
**Pull request by Nerev4r**
## About The Pull Request
There was supposed to be this funny little function where nanite slurry heals you based on the reac_volume of a touch application. It didn't prior because turns out adjustFireLoss and adjustBruteLoss don't work on synth limbs, thank you to @tf-4 for pointing this out for me and helping me get this off the ground. This now rebalances it, and it now works on both synths and organics.
The touch effect is gone, and nanite slurry now works as essentially a better omnizine for robotic parts. It has a strict overdose of twenty units, meaning it's only good for about ~40 brute or burn (Trust me, pure libital and aiuri go much faster and better, and are easily found in chem fridges.) After which, organics will vomit nanites on the floor (Some people might be familiar with that,) and synths will literally roast to death.
This also soups up System Cleaner a bit, due to the fact that it doesn't have a whole lot to do in the first place, and Nanite Slurry can be used as an incredibly effective synth poison with a hypospray or such.

## How This Contributes To The Skyrat Roleplay Experience
It'd be nice for synth-unique medicine to actually fix them. This way synths have literally anything to do with medical, and roboticists can maybe take advantage of plumbing. Also, Nanite Slurry can be used as essentially a poison for synths now.

## Changelog
:cl:
add: Nanite Slurry has been renovated to passively heal synths and the robolimbs of organics. Take caution to not have more than 20u in your body, though.
/:cl:

### CODE REVIEW

**Comment by Nerev4r** [comment]
Coming back to this soon, sware on me mum.

File: modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 1]
```diff
@@ -60,7 +60,7 @@
 
 
 /datum/reagent/medicine/nanite_slurry/on_mob_life(mob/living/carbon/M)
-	M.heal_bodypart_damage(0.8*REM, 0.8*REM, required_status = BODYPART_ROBOTIC)
+	M.heal_bodypart_damage(1*REM, 1*REM, required_status = BODYPART_ROBOTIC)
```
Comment (tf-4) [inline modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm]: make these vars on the chem actually

**Comment by Zergspower** [comment]
after dicking around with public chems I think this should be even higher - or have an additional one that does over X like organics have.  see below with this very easy to make 5u pill as my argument 
image

**Comment by tf-4** [comment]
> after dicking around with public chems I think this should be even higher - or have an additional one that does over X like organics have. see below with this very easy to make 5u pill as my argument

Firstly, trying to balance a race to be on par with humans is a fallacy. All races should be worse than humans. They are the baseline.
Secondly, balance should be changed in moderation. Make small, incremental buffs rather than large, sweeping buffs that then need large, sweeping nerfs two weeks down the line.
This is a fine change.

**Review by tf-4 — CHANGES_REQUESTED** [review]
Do put the healing as chem vars though. Hit me up on discord if you want a hand.

**Comment by Zergspower** [comment]
> 

Perhaps, but maybe raising the OD limit off of 20? for 20-40 max healing over a minute with 20u consumed is borderline pointless to use this chem at all, kind of why no one uses omni as a standalone.  combined with the increased time in self-heal would make getting overheated currently is a 100% death sentence since you wouldn't be able to out heal the burn. I agree with moderation but the changes we did in the other PR removed only a small percent off their increased damaged but also took away a chunk of their survivability

**Comment by Zergspower** [comment]
> > 
> 
> Perhaps, but maybe raising the OD limit off of 20? for 20-40 max healing over a minute with 20u consumed is borderline pointless to use this chem at all, kind of why no one uses omni as a standalone. combined with the increased time in self-heal would make getting overheated currently is a 100% death sentence since you wouldn't be able to out heal the burn. I agree with moderation but the changes we did in the other PR removed only a small percent off their increased damaged but also took away a chunk of their survivability

tweaking the recipe of the one i found above I can get 100/100 brute/burn healed in 32 seconds with 5u of SalAcid/Lenturi/Lib/Oxan - which is utterly overpowered given how easy it is to get them

File: modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 2]
```diff
@@ -57,10 +57,12 @@
 	metabolization_rate = 1.25 * REAGENTS_METABOLISM
 	process_flags = REAGENT_SYNTHETIC | REAGENT_ORGANIC
 	chemical_flags = REAGENT_CAN_BE_SYNTHESIZED
+	var/healing = 1
+	var/burning = 50
```
Comment (tf-4) [inline modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm]: auto-doc and the burning var isn't a great name imo
```suggestion
	/// How much brute and burn individually is healed per tick
	var/healing = 1
	/// How much body temperature is adjusted by per tick
	var/temperature_change = 50
```

File: modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 3]
```diff
@@ -50,21 +50,30 @@
...
-/datum/reagent/medicine/nanite_slurry/expose_mob(mob/living/M, method=TOUCH, reac_volume, show_message = 1)
-	if(iscarbon(M))
-		if(!(method in list(INGEST, VAPOR, INJECT)))
-			M.adjustFireLoss(-reac_volume)
-			M.adjustBruteLoss(-reac_volume)
-			if(show_message)
-				to_chat(M, span_notice("You feel much better..."))
-	..()
 
 /datum/reagent/medicine/nanite_slurry/on_mob_life(mob/living/carbon/M)
```
Comment (tf-4) [inline modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm]: also rename this (and all the others in this proc) to something more descriptive
```suggestion
/datum/reagent/medicine/nanite_slurry/on_mob_life(mob/living/carbon/affected_mob)
```

File: modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 4]
```diff
@@ -50,21 +50,30 @@
...
-	M.heal_bodypart_damage(0.5*REM, 0.5*REM, required_status = BODYPART_ROBOTIC)
+	M.heal_bodypart_damage(healing*REM, healing*REM, required_status = BODYPART_ROBOTIC)
+	..()
+	. = 1
+
+/datum/reagent/medicine/nanite_slurry/overdose_process(mob/living/carbon/M, delta_time, times_fired)
```
Comment (tf-4) [inline modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm]: same here
```suggestion
/datum/reagent/medicine/nanite_slurry/overdose_process(mob/living/carbon/affected_mob, delta_time, times_fired)
```

**Comment by RatFromTheJungle** [comment]
> > > 
> > 
> > 
> > Perhaps, but maybe raising the OD limit off of 20? for 20-40 max healing over a minute with 20u consumed is borderline pointless to use this chem at all, kind of why no one uses omni as a standalone. combined with the increased time in self-heal would make getting overheated currently is a 100% death sentence since you wouldn't be able to out heal the burn. I agree with moderation but the changes we did in the other PR removed only a small percent off their increased damaged but also took away a chunk of their survivability
> 
> tweaking the recipe of the one i found above I can get 100/100 brute/burn healed in 32 seconds with 5u of SalAcid/Lenturi/Lib/Oxan - which is utterly overpowered given how easy it is to get them

see: above

They don't need to be comparable to humans, they have a ton of enviromental protections and immunity too **A LOT** of antag weapons, they can deal with healing slower.

File: modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 5]
```diff
@@ -50,21 +50,32 @@
 
 /datum/reagent/medicine/nanite_slurry
 	name = "Nanite Slurry"
-	description = "If used in touch-based applications, immediately repairs and refurbishes synthetic lifeforms, also does that while circulating in their system."
+	description = "A localized swarm of nanomachines specialized in repairing mechanical parts. Due to the nanites needing to interface with the host's systems to repair them, a surplus of them will cause them to overheat, or for the swarm to eject out of the mouth for safety."
```
Comment (tf-4) [inline modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm]: Do IPCs have mouths? I don't want to merge failRP here...

File: modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 6]
```diff
@@ -50,21 +50,32 @@
...
-				to_chat(M, span_notice("You feel much better..."))
+/datum/reagent/medicine/nanite_slurry/on_mob_life(mob/living/carbon/affected_mob)
+	affected_mob.heal_bodypart_damage(healing*REM, healing*REM, required_status = BODYPART_ROBOTIC)
 	..()
+	. = 1
 
-/datum/reagent/medicine/nanite_slurry/on_mob_life(mob/living/carbon/M)
-	M.heal_bodypart_damage(0.5*REM, 0.5*REM, required_status = BODYPART_ROBOTIC)
+/datum/reagent/medicine/nanite_slurry/overdose_process(mob/living/carbon/affected_mob, delta_time, times_fired)
+	if(!(affected_mob.mob_biotypes & MOB_ROBOTIC))
```
Comment (tf-4) [inline modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm]: Early return this instead

**Review by tf-4 — CHANGES_REQUESTED** [review]
Last few bits code-wise

File: modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 7]
```diff
@@ -50,21 +50,30 @@
...
-	if(iscarbon(M))
-		if(!(method in list(INGEST, VAPOR, INJECT)))
-			M.adjustFireLoss(-reac_volume)
-			M.adjustBruteLoss(-reac_volume)
-			if(show_message)
-				to_chat(M, span_notice("You feel much better..."))
+/datum/reagent/medicine/nanite_slurry/on_mob_life(mob/living/carbon/affected_mob)
+	affected_mob.heal_bodypart_damage(healing*REM, healing*REM, required_status = BODYPART_ROBOTIC)
 	..()
+	. = 1
```
Comment (tf-4) [inline modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm]:
```suggestion
```

**Review by tf-4 — APPROVED** [review]

File: code/__DEFINES/misc.dm [hunk 8]
```diff
@@ -499,6 +499,7 @@ GLOBAL_LIST_INIT(pda_styles, sort_list(list(MONO, VT, ORBITRON, SHARE)))
 
 #define VOMIT_TOXIC 1
 #define VOMIT_PURPLE 2
+#define VOMIT_NANITE 3 //SKYRAT EDIT ADD
```
Comment (GoldenAlpharex) [inline code/__DEFINES/misc.dm]: You should say in your SKYRAT EDIT comment the reason why you're making said edit, so it's easier to find in the future.

File: code/game/turfs/turf.dm [hunk 9]
```diff
@@ -617,6 +617,10 @@ GLOBAL_LIST_EMPTY(station_turfs)
 		V.icon_state = "vomitpurp_[pick(1,4)]"
 	else if (toxvomit == VOMIT_TOXIC)
 		V.icon_state = "vomittox_[pick(1,4)]"
+	else if (toxvomit == VOMIT_NANITE) //SKYRAT EDIT BEGIN
+		V.name = "metallic slurry"
+		V.desc = "A puddle of metallic slurry that looks vaguely like very fine sand. It almost seems like it's moving..."
+		V.icon_state = "vomitnanite_[pick(1,4)]" // SKYRAT EDIT END
```
Comment (GoldenAlpharex) [inline code/game/turfs/turf.dm]:
```suggestion
	//SKYRAT EDIT START - Nanite Slurry
	else if (toxvomit == VOMIT_NANITE)
		V.name = "metallic slurry"
		V.desc = "A puddle of metallic slurry that looks vaguely like very fine sand. It almost seems like it's moving..."
		V.icon_state = "vomitnanite_[pick(1,4)]"
	// SKYRAT EDIT END
```

File: modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 10]
```diff
@@ -23,16 +23,16 @@
 	reagent_state = LIQUID
 	color = "#F1C40F"
 	taste_description = "ethanol"
-	metabolization_rate = 0.5 * REAGENTS_METABOLISM
+	metabolization_rate = 2 * REAGENTS_METABOLISM
 	process_flags = REAGENT_SYNTHETIC
 
-/datum/reagent/medicine/system_cleaner/on_mob_life(mob/living/carbon/M)
+/datum/reagent/medicine/system_cleaner/on_mob_life(mob/living/carbon/M, delta_time, times_fired)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm]:
```suggestion
/datum/reagent/medicine/system_cleaner/on_mob_life(mob/living/carbon/affected_mob, delta_time, times_fired)
```

File: modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 11]
```diff
@@ -23,16 +23,16 @@
...
 	M.adjustToxLoss(-2*REM, 0)
 	. = 1
 	for(var/A in M.reagents.reagent_list)
 		var/datum/reagent/R = A
 		if(R != src)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm]:
```suggestion
	affected_mob.adjustToxLoss(-2 * REM, 0)
	. = 1
	for(var/thing in affected_mob.reagents.reagent_list)
		var/datum/reagent/reagent = thing
		if(reagent != src)
```

File: modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 12]
```diff
@@ -23,16 +23,16 @@
...
-			M.reagents.remove_reagent(R.type,1)
+			M.reagents.remove_reagent(R.type, 1 * REM * delta_time)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm]:
```suggestion
			affected_mob.reagents.remove_reagent(reagent.type, 1 * REM * delta_time)
```

File: modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 13]
```diff
@@ -50,21 +50,29 @@
...
 
-/datum/reagent/medicine/nanite_slurry/expose_mob(mob/living/M, method=TOUCH, reac_volume, show_message = 1)
-	if(iscarbon(M))
-		if(!(method in list(INGEST, VAPOR, INJECT)))
-			M.adjustFireLoss(-reac_volume)
-			M.adjustBruteLoss(-reac_volume)
-			if(show_message)
-				to_chat(M, span_notice("You feel much better..."))
+/datum/reagent/medicine/nanite_slurry/on_mob_life(mob/living/carbon/affected_mob)
+	affected_mob.heal_bodypart_damage(healing*REM, healing*REM, required_status = BODYPART_ROBOTIC)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm]:
```suggestion
	affected_mob.heal_bodypart_damage(healing * REM, healing * REM, required_status = BODYPART_ROBOTIC)
```

File: modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 14]
```diff
@@ -50,21 +50,29 @@
...
 	..()
 
-/datum/reagent/medicine/nanite_slurry/on_mob_life(mob/living/carbon/M)
-	M.heal_bodypart_damage(0.5*REM, 0.5*REM, required_status = BODYPART_ROBOTIC)
+/datum/reagent/medicine/nanite_slurry/overdose_process(mob/living/carbon/affected_mob, delta_time, times_fired)
+	if(affected_mob.mob_biotypes & MOB_ROBOTIC)
+		affected_mob.adjust_bodytemperature(temperature_change * REM * delta_time)
+		return ..()
+	affected_mob.reagents.remove_reagent(type, 3.6) //gets removed from organics very fast
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm]: That 3.6 should be a define, I have no idea why it'd be that and not, say, 4, and it just in generally doesn't mean anything to me.

File: modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 15]
```diff
@@ -50,21 +50,29 @@
...
+	if(prob(25))
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm]: This also should be a define, like `#define NANITE_SLURRY_ORGANIC_VOMIT_CHANCE 25`

**Review by GoldenAlpharex — CHANGES_REQUESTED** [review]
That sounds kinda useful, although you should change the PR description to reflect it's `1` and not `0.8` (unless this is working in even weirder ways than I'm assuming it does).

Also, I'd argue at this point it's kind of not just a fix.

**Review by GoldenAlpharex — COMMENTED** [review]
Those changes are good, I'll check if anything's missing.

File: code/__DEFINES/misc.dm (hunk 8 above)
Comment (GoldenAlpharex) [reply code/__DEFINES/misc.dm]:
```suggestion
#define VOMIT_NANITE 3 //SKYRAT EDIT ADD - Nanite Slurry
```

File: modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 16]
```diff
@@ -50,21 +53,32 @@
...
 	reagent_state = LIQUID
 	color = "#cccccc"
-	process_flags = REAGENT_SYNTHETIC
+	overdose_threshold = 20
+	metabolization_rate = 1.25 * REAGENTS_METABOLISM
+	process_flags = REAGENT_SYNTHETIC | REAGENT_ORGANIC
+	chemical_flags = REAGENT_CAN_BE_SYNTHESIZED
+	/// How much brute and burn individually is healed per tick
+	var/healing = 1
+	/// How much body temperature is adjusted by per tick
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/customization/modules/reagents/chemistry/reagents/medicine_reagents.dm]:
```suggestion
	/// How much body temperature is increased by per overdose cycle on robotic bodyparts.
```

**Review by GoldenAlpharex — APPROVED** [review]
We good to go.

**Pull request merged**